design-documents/backward-compatibility/api-vs-spi.md

```diff
@@ -0,0 +1,78 @@
+This makes those classes stuck in time.
+It is impossible to add new methods to the class because it is a breaking change for an SPI class, and it is impossible to remove methods because it is a breaking change for an API class.
+
+From another side, Magento [Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html) discourage usage of inheritance ("2.6. Inheritance SHOULD NOT be used. Composition SHOULD be used for code reuse.").
+This means that only interfaces should be considered SPI, while classes should be considered API-only code.  
+
+## Problem Statement
+
+As we're trying to find the balance between keeping extensions as compatible with core as possible, while allowing core developers to make some necessary changes, [Codebase changes](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/versioning/codebase-changes.html) and [Module version dependencies](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/versioning/dependencies.html) describe more and more rules and exceptions.
+Also, see "classes explicitly meant for extension" in [Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html?itm_source=devdocs&itm_medium=quick_search&itm_campaign=federated_search&itm_term=defin#review).
+It becomes hard to remember and follow all the described recommendations.
+It is also more difficult to follow a separate document than codebase itself.
+And `@api` annotation used in the codebase is not sufficient now to guide a developer about dependencies.
+
+## Solution
+
+Distinguish API and SPI using different annotation in the code:
+
+* `@api` for API
+* `@spi` for SPI
+
+This simplifies rules and helps get rid of exceptional cases.
+See [updated Dependencies page](https://github.com/magento/devdocs/pull/5085/files) for possible simplifications of guidelines.
+
+Guidelines should describe:
+* API
+   * changes in API increase MAJOR version of the module
+   * extension developers should depend on MAJOR version if they call API methods
+* SPI
+   * changes in SPI increase MINOR version of the module 
+   * extension developers should depend on MINOR version if they implement/extend SPI
+   * XML configuration files are SPI. Special cases for `di.xml`
+      * virtual types are private code by default. Can be marked as `@api` and/or `@spi`
+      * lists (e.g., `CommandList`) can be marked as `@spi`
+   * all classes intended for extension are SPI
+ 
+## Scope of Work
+
+### Initial Demarcation
+
+Initially, mark the code as follows:
+
+* For all interfaces marked with `@api`, also add `@spi`. Interfaces are intended for implementation, so they are SPI. Most of the interfaces are also intended for usage, so they are API. There might be exceptions, but those can be fixed in the future or used for new interfaces.
```

Comment:
Just my opinion, to have both `@api` and `@spi` annotations for the same interface/class looks more complicated at least because they have different versioning rules.